### PR TITLE
Use ejs template engine to render API results

### PR DESCRIPTION
**Pull request by kirgene**

### CODE REVIEW

**Review by bushjames — APPROVED** [review]
I cant spot any issues. Many thanks for expanding the capability of this service Yev.

File: package.json [hunk 1]
```diff
@@ -10,37 +10,56 @@
...
-    "object.fromentries": "github:es-shims/Object.fromEntries#v2.0.2",
-    "string.prototype.matchall": "github:es-shims/String.prototype.matchAll#v4.0.2"
+    "koa": "^2.13.1",
+    "koa-sendfile": "^3.0.0",
+    "mysql2": "^2.2.5",
+    "node-table-to-csv": "^0.1.0",
+    "object.fromentries": "github:es-shims/Object.fromEntries#v2.0.4",
+    "patch-package": "^6.4.7",
+    "puppeteer": "^8.0.0",
+    "string.prototype.matchall": "github:es-shims/String.prototype.matchAll#v4.0.4"
```
Comment (partiallyordered) [inline package.json:63]: Looks like this is no longer used as a result of these changes
```suggestion
    "string.prototype.matchall": "github:es-shims/String.prototype.matchAll#v4.0.4"
```

File: src/handlers.js [hunk 2]
```diff
@@ -98,22 +82,23 @@ const createReportHandlers = (reportsConfig) => {
...
-                ctx.state.logger.push({ dbQuery, queryArgs }).log('Executing query');
-                const result = await ctx.db.query(dbQuery, queryArgs);
-                ctx.state.logger.push({ dbQuery, queryArgs, result }).log('Query result');
 
-                ctx.response.body = result;
+                const queries = Object.entries(template.dataSource.data)
+                    .map(([k, v]) => ctx.db.query(v, queryArgs)
+                        .then((result) => ({ [k]: result })));
+                const result = await Promise.all(queries);
+                ctx.response.html = template.render(Object.assign({}, ...result));
```
Comment (partiallyordered) [inline src/handlers.js:90]: I suspect probably this should go on `ctx.state.html` or similar. `ctx.state` should be reset per-request, and `ctx.response.html`, while not part of the API, could _possibly_ be part of the API in future.

Specifically, `ctx.state` is intended for this kind of data passing.

File: src/handlers.js [hunk 3]
```diff
@@ -98,22 +82,23 @@ const createReportHandlers = (reportsConfig) => {
...
             },
         };
-        return [route, handler];
+        return [`/${route}`, handler];
     };
 
-    return fromEntries(Object.entries(reportsConfig).map(createReportHandler));
+    return fromEntries(Object.entries(reportTemplates).map(createReportHandler));
```
Comment (partiallyordered) [inline src/handlers.js:96]: `fromEntries` is probably actually available now as `Object.fromEntries`. I guess the shim will use that if it's available, so no huge rush to get rid of it if you can't be bothered right now.

File: templates/participants.yaml [hunk 4]
```diff
@@ -0,0 +1,29 @@
+params:
+  name:
+    required: false
+    default: null
+  currency:
+    required: true
+  limit:
+    required: false
+    default: 100
```
Comment (partiallyordered) [inline templates/participants.yaml:9]: This might be something of a surprising default. @bushjames how many participants for our current project?

**Review by partiallyordered — COMMENTED** [review]
Some general comments:
- Could we smoke-check the chart in CI with `helm template somename helm/reporting-service | kubeval --strict`? Example using GH actions here: https://github.com/modusintegration/finance-portal-settlements/blob/master/.github/workflows/validate-chart.yml could be reused here easily (and should be fine alongside the existing CI)
- Is there a reason this chart isn't in the the mojaloop/helm GH repo?
- Why create a service account in the chart?
- Does the helm hook to test the pod connection do anything that a readiness/liveness probe wouldn't/shouldn't? Is the added dependency on helm hooks, and the increased complexity worthwhile?

File: Dockerfile [hunk 5]
```diff
@@ -1,18 +1,23 @@
-FROM node:12.18.0-alpine as builder
+FROM node:lts-buster-slim as builder
```
[the remark this answers is not in the document]
Comment (kirgene) [reply Dockerfile:1]: We are using Puppeteer here and it won't work with musl libc provided by alpine

File: package.json [hunk 6]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "mojaloop-reporting-service",
-  "version": "10.4.1",
+  "version": "10.4.5",
```
[the remark this answers is not in the document]
Comment (kirgene) [reply package.json:3]: There were several "internal" versions of that service that I tested and I was afraid I could have issues with CI/CD pipeline picking old cached stuff.
So should I modify it to `10.4.2`?

[the remark this answers is not in the document]
Comment (partiallyordered) [reply package.json:3]: Eh, it's 10.4.5 in master now, it doesn't really matter.

File: templates/participants.yaml (hunk 4 above)
Comment (kirgene) [reply templates/participants.yaml:9]: this template config exists for example/documentation purposes and won't be used in prod.

File: src/handlers.js (hunk 3 above)
Comment (kirgene) [reply src/handlers.js:96]: done in PR https://github.com/mojaloop/reporting/pull/17

File: src/handlers.js (hunk 2 above)
Comment (kirgene) [reply src/handlers.js:90]: done in PR https://github.com/mojaloop/reporting/pull/17

File: package.json (hunk 1 above)
Comment (kirgene) [reply package.json:63]: done in PR https://github.com/mojaloop/reporting/pull/17

File: package.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (kirgene) [reply package.json:3]: done in PR https://github.com/mojaloop/reporting/pull/17

File: README.md [hunk 7]
```diff
@@ -56,18 +44,13 @@ docker run -v $PWD/config:/opt/reporting/config -p 3000:3000 --env-file=./.my.en
  3. If an issue must be ignored, and **it is absolutely safe to do so**, run `npm run audit:resolve` and select "remind me in 24h"
 
 #### TODO
-- Probably get rid of knex in favour of plain mysql2?
 - The initial implementation was developed with compatibility for Jaspersoft Studio queries in
     mind. Optionally use different templating engines. Parametrise this in the config. Perhaps a
     single default global templating engine, and then a per-report override.
```
[the remark this answers is not in the document]
Comment (kirgene) [reply README.md:49]: done in PR https://github.com/mojaloop/reporting/pull/17

File: .eslintrc.json [hunk 8]
```diff
@@ -8,6 +8,7 @@
       "error",
       4,
       { "SwitchCase": 1 }
-    ]
+    ],
+    "no-restricted-syntax": "off"
```
[the remark this answers is not in the document]
Comment (kirgene) [reply .eslintrc.json:12]: seems like an unnecessary rule already, fixed in PR https://github.com/mojaloop/reporting/pull/17

File: helm/reporting-service/templates/secret.yaml [hunk 9]
```diff
@@ -0,0 +1,7 @@
+apiVersion: v1
+kind: Secret
+metadata:
+  name: {{ .Release.Name }}-auth
```
[the remark this answers is not in the document]
Comment (kirgene) [reply helm/reporting-service/templates/secret.yaml:4]: done in PR #17

**Pull request merged**